diff --git a/src/renderer/components/ScreenSharePicker.tsx b/src/renderer/components/ScreenSharePicker.tsx
index db3c483..81929b0 100644
--- a/src/renderer/components/ScreenSharePicker.tsx
+++ b/src/renderer/components/ScreenSharePicker.tsx
@@ -34,6 +34,7 @@ import { onceReady } from "@equicord/types/webpack";
 import { FluxDispatcher, MediaEngineStore, Select, UserStore, useState } from "@equicord/types/webpack/common";
 import { Node } from "@vencord/venmic";
 import type { Dispatch, SetStateAction } from "react";
+import { patchOverrideDevices } from "renderer/patches/screenShareFixes";
 import { State, useSettings, useVesktopState } from "renderer/settings";
 import { isLinux, isWindows } from "renderer/utils";
 
@@ -59,6 +60,7 @@ interface AudioItem {
 
 interface StreamSettings {
     audio: boolean;
+    overrideVideoDevice?: string;
     contentHint?: string;
     includeSources?: AudioSources;
     excludeSources?: AudioSources;
@@ -123,6 +125,10 @@ export function openScreenSharePicker(screens: Source[], skipPicker: boolean) {
                             }
                         }
 
+                        patchOverrideDevices({
+                            video: v.overrideVideoDevice
+                        });
+
                         resolve(v);
                     }}
                     close={() => {
@@ -330,6 +336,11 @@ function StreamSettingsUi({
         }
     );
 
+    const [videoDevices, , videoDevicesPending] = useAwaiter(
+        () => navigator.mediaDevices.enumerateDevices().then(g => g.filter(d => d.kind === "videoinput")),
+        { fallbackValue: [] }
+    );
+
     const openSettings = () => {
         openModal(props => (
             <AudioSettingsModal
@@ -391,6 +402,7 @@ function StreamSettingsUi({
                                 a much sharper and clearer image.
                             </Paragraph>
                         </div>
+
                         {isWindows && (
                             <FormSwitch
                                 title="Stream With Audio"
@@ -403,6 +415,26 @@ function StreamSettingsUi({
                     </section>
                 </div>
 
+                <div>
+                    <Heading tag="h5">
+                        {videoDevicesPending ? "Loading video devices..." : "Video devices"}
+                    </Heading>
+                    <Select
+                        options={videoDevices.map(({ label, deviceId }) => ({
+                            label,
+                            value: deviceId
+                        }))}
+                        isSelected={d => settings.overrideVideoDevice === d}
+                        select={d => {
+                            setSettings(v => ({ ...v, overrideVideoDevice: d }));
+                        }}
+                        serialize={String}
+                        popoutPosition="top"
+                        closeOnSelect
+                        isDisabled={videoDevicesPending}
+                    />
+                </div>
+
                 {isLinux && (
                     <AudioSourcePickerLinux
                         openSettings={openSettings}
diff --git a/src/renderer/patches/screenShareFixes.ts b/src/renderer/patches/screenShareFixes.ts
index 0769771..ab29687 100644
--- a/src/renderer/patches/screenShareFixes.ts
+++ b/src/renderer/patches/screenShareFixes.ts
@@ -11,21 +11,49 @@ import { isLinux } from "renderer/utils";
 
 const logger = new Logger("EquibopStreamFixes");
 
-if (isLinux) {
-    const original = navigator.mediaDevices.getDisplayMedia;
-
-    async function getVirtmic() {
-        try {
-            const devices = await navigator.mediaDevices.enumerateDevices();
-            const audioDevice = devices.find(({ label }) => label === "vencord-screen-share");
-            return audioDevice?.deviceId;
-        } catch (error) {
-            return null;
-        }
+const original = navigator.mediaDevices.getDisplayMedia;
+
+interface OverrideDevices {
+    video: string | undefined;
+}
+
+let overrideDevices: OverrideDevices = { video: undefined };
+
+export const patchOverrideDevices = (newOverrideDevices: OverrideDevices) => {
+    overrideDevices = newOverrideDevices;
+};
+
+async function getVirtmic() {
+    try {
+        const devices = await navigator.mediaDevices.enumerateDevices();
+        const audioDevice = devices.find(({ label }) => label === "vencord-screen-share");
+        return audioDevice?.deviceId;
+    } catch (error) {
+        return null;
+    }
+}
+
+navigator.mediaDevices.getDisplayMedia = async function (opts) {
+    const stream = await original.call(this, opts);
+
+    if (overrideDevices.video) {
+        const video = await navigator.mediaDevices.getUserMedia({
+            video: {
+                deviceId: { exact: overrideDevices.video }
+            }
+        });
+
+        stream.getVideoTracks().forEach(t => {
+            t.stop();
+            stream.removeTrack(t);
+        });
+
+        video.getVideoTracks().forEach(t => {
+            stream.addTrack(t);
+        });
     }
 
-    navigator.mediaDevices.getDisplayMedia = async function (opts) {
-        const stream = await original.call(this, opts);
+    if (isLinux) {
         const id = await getVirtmic();
 
         const frameRate = Number(State.store.screenshareQuality?.frameRate ?? 30);
@@ -69,7 +97,7 @@ if (isLinux) {
             stream.getAudioTracks().forEach(t => stream.removeTrack(t));
             stream.addTrack(audio.getAudioTracks()[0]);
         }
+    }
 
-        return stream;
-    };
-}
+    return stream;
+};
